wandb/sdk/internal/job_builder.py

```diff
@@ -0,0 +1,252 @@
+        self._used_job = val
+
+    def _set_logged_code_artifact(
+        self, res: Optional[Dict], artifact: "ArtifactRecord"
+    ) -> None:
+        if artifact.type == "code" and res is not None:
+            self._logged_code_artifact = ArtifactInfoForJob(
+                {
+                    "id": res["id"],
+                    "name": artifact.name,
+                }
+            )
+
+    def _build_repo_job(
+        self, remote: str, commit: str, program_relpath: str, args: List[str]
+    ) -> Tuple[Artifact, GitSourceDict]:
+        source: GitSourceDict = {
+            "entrypoint": [
+                os.path.basename(sys.executable),
+                program_relpath,
+            ],
+            "args": args,
+            "git": {
+                "remote": remote,
+                "commit": commit,
+            },
+        }
+
+        name = make_artifact_name_safe(f"job-{remote}_{program_relpath}")
+
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        if os.path.exists(os.path.join(self._settings.files_dir, DIFF_FNAME)):
+            artifact.add_file(
+                os.path.join(self._settings.files_dir, DIFF_FNAME),
+                name="diff.patch",
+            )
+        return artifact, source
+
+    def _build_artifact_job(
+        self, program_relpath: str, args: List[str]
+    ) -> Tuple[Artifact, ArtifactSourceDict]:
+        assert isinstance(self._logged_code_artifact, dict)
+        source: ArtifactSourceDict = {
+            "entrypoint": [
+                os.path.basename(sys.executable),
+                program_relpath,
+            ],
+            "args": args,
+            "artifact": f"wandb-artifact://_id/{self._logged_code_artifact['id']}",
+        }
+
+        name = f"job-{self._logged_code_artifact['name']}"
+
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        return artifact, source
+
+    def _build_image_job(
+        self, image_name: str, args: List[str]
+    ) -> Tuple[Artifact, ImageSourceDict]:
+        name = make_artifact_name_safe(f"job-{image_name}")
+        artifact = Artifact(name, JOB_ARTIFACT_TYPE)
+        source: ImageSourceDict = {
+            "image": image_name,
+            "args": args,
+        }
+        return artifact, source
+
+    def build(self) -> Optional[Artifact]:
+        if not os.path.exists(
+            os.path.join(self._settings.files_dir, REQUIREMENTS_FNAME)
+        ):
+            return None
+        metadata = self._handle_metadata_file()
+        if metadata is None:
+            return None
+        git_info: Dict[str, str] = metadata.get("git", {})
+        image_name: Optional[str] = metadata.get("docker", None)
+        program_relpath: Optional[str] = metadata.get("codePath", None)
+        args: List[str] = metadata.get("args", [])
+        runtime: Optional[str] = metadata.get("python", None)
+        # can't build a job without a python version
+        if runtime is None:
+            return None
```

Comment:
nit: should you move after the metadata check since if it is None, you don't need to extract the other fields?

Reply:
nice, fixed